### ISSUE TITLE
PL/Python KeyError issue with set returning functions

### ISSUE THREAD

**Issue by 0x0FFF** [greenplum-db/gpdb#492]
The issue can be reproduced with this simple code:

```
CREATE OR REPLACE FUNCTION func1(var_cause_bug int4[]) RETURNS SETOF int4 AS $$
for el in var_cause_bug:
    yield el
$$ LANGUAGE plpythonu;

CREATE OR REPLACE FUNCTION func2() RETURNS int4 AS $$
return 1
$$ LANGUAGE plpythonu;

select func1(array[1,2,3]), func1(array[1,2,3]);
select func2();
```

You would see the following output:

```
vagrant=# select func1(array[1,2,3]), func1(array[1,2,3]);
 func1 | func1 
-------+-------
     1 |     1
     2 |     2
     3 |     3
(3 rows)

vagrant=# select func2();
ERROR:  could not compile PL/Python function "func2" (plpython.c:4648)
DETAIL:  KeyError: 'var_cause_bug'
```

**Comment by 0x0FFF** [comment]
Root cause of the issue is the function `PLy_function_delete_args` which makes `PyDict_DelItemString` without checking whether the item really exists in global dictionary or not

**Comment by hlinnaka** [comment]
pgsql-hackers discussion: https://www.postgresql.org/message-id/flat/7776.1459882802%40sss.pgh.pa.us#7776.1459882802@sss.pgh.pa.us

The upstream commit that resulted: 7e3bb080387f4143cdc908bf97daf9a8abdc445f.

I believe what remains TODO here is to sync up the code that was committed to GPDB, with the version that went into upstream. I believe it was not identical, and at the end of that pgsql-hackers thread, Tom referred to some cases that the original patch version didn't fix.

**Comment by BaiShaoqi** [comment]
This was fixed and committed to GPDB in d54de3962a12e6075eae80fe75afc3ab59eea0d5
